File: packages/transport/rpc_dart_wasm/ios/Classes/RpcDartWasmPlugin.swift
import Flutter
import WebKit

public class RpcDartWasmPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var messenger: FlutterBinaryMessenger?
    private var runtimes: [String: WasmRuntime] = [:]
    private var checkSupportWebView: WKWebView?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "rpc_dart_wasm",
            binaryMessenger: registrar.messenger()
        )
        let instance = RpcDartWasmPlugin()
        instance.channel = channel
        instance.messenger = registrar.messenger()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "checkSupport":
            checkSupport(result: result)
        case "loadRuntime":
            let args = call.arguments as? [String: Any]
            let wasmData = (args?["wasm"] as? FlutterStandardTypedData)?.data
            let mjsCode = args?["mjs"] as? String
            let prefix = args?["jsBootPrefix"] as? String ?? ""
            loadRuntime(wasmBytes: wasmData, mjsCode: mjsCode, jsBootPrefix: prefix, result: result)
        case "closeRuntime":
            let args = call.arguments as? [String: Any]
            closeRuntime(runtimeId: args?["runtimeId"] as? String ?? "")
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Check support

    private func checkSupport(result: @escaping FlutterResult) {
        let webView = WKWebView(frame: .zero)
        checkSupportWebView = webView
        webView.evaluateJavaScript("""
            (function() {
              var r = {};
              r.jsEngineAvailable = true;
              r.hasWebAssembly = (typeof WebAssembly !== 'undefined');
              if (r.hasWebAssembly) {
                try {
                  var bytes = new Uint8Array([
                    0x00, 0x61, 0x73, 0x6D, 0x01, 0x00, 0x00, 0x00,
                    0x01, 0x05, 0x01, 0x5F, 0x01, 0x7F, 0x01
                  ]);
                  r.wasmGC = WebAssembly.validate(bytes) ? 'GC_SUPPORTED' : 'GC_NOT_VALIDATED';
                } catch(e) { r.wasmGC = 'GC_ERROR: ' + e; }
              }
              return JSON.stringify(r);
            })()
        """) { [weak self] value, error in
            self?.checkSupportWebView = nil
            if let json = value as? String,
               let data = json.data(using: .utf8),
               let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                result(dict)
            } else {
                result([
                    "jsEngineAvailable": true,
                    "hasWebAssembly": false,
                    "error": error?.localizedDescription ?? "unknown"
                ] as [String: Any])
            }
        }
    }

    // MARK: - Load runtime

    private func loadRuntime(
        wasmBytes: Data?,
        mjsCode: String?,
        jsBootPrefix: String,
        result: @escaping FlutterResult
    ) {
        guard let wasmBytes = wasmBytes, let mjsCode = mjsCode else {
            result(["error": "Missing wasm or mjs data"])
            return
        }
        guard let messenger = self.messenger else {
            result(["error": "No binary messenger"])
            return
        }

        let runtimeId = UUID().uuidString
        let runtime = WasmRuntime(
            runtimeId: runtimeId,
            messenger: messenger,
            rxChannel: runtimeRxChannel(runtimeId),
            wasmBytes: wasmBytes
        )
        runtimes[runtimeId] = runtime
        registerByteChannel(runtimeId: runtimeId)

        let plainJs = stripModuleSyntax(mjsCode)

        let bootHtml = """
        <!DOCTYPE html><html><head><meta charset="utf-8"></head><body><script>
        var _nativeSetTimeout = setTimeout;
        var _nativeClearTimeout = clearTimeout;
        var _microtaskQueue = [];
        var _timerId = 0;
        var _timers = {};
        var _nextTickHandle = 0;
        function queueMicrotask(fn) { _microtaskQueue.push(fn); }
        function _flushMicrotasks() {
          while (_microtaskQueue.length > 0) {
            var fn = _microtaskQueue.shift();
            try { fn(); } catch(e) { console.error(e); }
          }
        }
        var _origConsole = typeof console !== 'undefined' ? console : {};
        console = {
          log: function() { var m = Array.prototype.join.call(arguments, ' '); window.webkit.messageHandlers.rpcConsole.postMessage({level: 'info', message: m}); },
          warn: function() { var m = Array.prototype.join.call(arguments, ' '); window.webkit.messageHandlers.rpcConsole.postMessage({level: 'warn', message: m}); },
          error: function() { var m = Array.prototype.join.call(arguments, ' '); window.webkit.messageHandlers.rpcConsole.postMessage({level: 'error', message: m}); },
          info: function() { var m = Array.prototype.join.call(arguments, ' '); window.webkit.messageHandlers.rpcConsole.postMessage({level: 'info', message: m}); },
          debug: function() { var m = Array.prototype.join.call(arguments, ' '); window.webkit.messageHandlers.rpcConsole.postMessage({level: 'debug', message: m}); }
        };
        function setTimeout(fn, ms) {
          var id = ++_timerId;
          _timers[id] = { fn: fn, interval: false, ms: ms || 0, next: Date.now() + (ms || 0) };
          _scheduleNextTick();
          return id;
        }
        function setInterval(fn, ms) {
          var id = ++_timerId;
          _timers[id] = { fn: fn, interval: true, ms: ms || 16, next: Date.now() + (ms || 16) };
          _scheduleNextTick();
          return id;
        }
        function clearInterval(id) { delete _timers[id]; }
        function clearTimeout(id) { delete _timers[id]; }
        function _scheduleNextTick() {
          var now = Date.now();
          var minDelay = Infinity;
          var ids = Object.keys(_timers);
          for (var i = 0; i < ids.length; i++) {
            var t = _timers[ids[i]];
            if (t) {
              var d = t.next - now;
              if (d < minDelay) minDelay = d;
            }
          }
          if (_nextTickHandle) {
            _nativeClearTimeout(_nextTickHandle);
            _nextTickHandle = 0;
          }
          if (minDelay === Infinity) return;
          if (minDelay < 0) minDelay = 0;
          _nextTickHandle = _nativeSetTimeout(_runDueTimers, minDelay);
        }
        function _runDueTimers() {
          _nextTickHandle = 0;
          var now = Date.now();
          var ids = Object.keys(_timers);
          for (var i = 0; i < ids.length; i++) {
            var id = ids[i];
            var t = _timers[id];
            if (!t) continue;
            if (now >= t.next) {
              if (t.interval) {
                t.next = now + t.ms;
              } else {
                delete _timers[id];
              }
              try { t.fn(); } catch(e) { console.error(e); }
            }
          }
          _flushMicrotasks();
          _scheduleNextTick();
        }
        function _rpcWasmSendBytes(bytes) {
          fetch('rpc-wasm:///send', {method: 'POST', body: bytes});
        }
        var _recvRunning = false;
        function _rpcWasmReceiveBytes(bytes) {
          if (typeof rpcWasmReceiveBytes === 'function') {
            rpcWasmReceiveBytes(bytes);
          } else if (typeof globalThis.rpcWasmReceiveBytes === 'function') {
            globalThis.rpcWasmReceiveBytes(bytes);
          }
          _flushMicrotasks();
        }
        function _startRecvLoop() {
          if (_recvRunning) return;
          _recvRunning = true;
          (function poll() {
            fetch('rpc-wasm:///recv').then(function(r) {
              return r.arrayBuffer();
            }).then(function(buf) {
              _rpcWasmReceiveBytes(new Uint8Array(buf));
              poll();
            }).catch(function(e) {
              _recvRunning = false;
            });
          })();
        }
        \(jsBootPrefix)
        \(plainJs)
        (async function() {
          try {
            var resp = await fetch('rpc-wasm:///module.wasm');
            var compiled;
            if (typeof compileStreaming === 'function') {
              compiled = await compileStreaming(resp);
            } else {
              var wasmBytes = new Uint8Array(await resp.arrayBuffer());
              compiled = await compile(wasmBytes);
            }
            var app = await compiled.instantiate({});
            _startRecvLoop();
            app.invokeMain();
            _flushMicrotasks();
            window.webkit.messageHandlers.rpcBoot.postMessage('ok');
          } catch(e) {
            window.webkit.messageHandlers.rpcBoot.postMessage('error:' + e + (e && e.stack ? '\\n' + e.stack : ''));
          }
        })();
        </script></body></html>
        """

        runtime.boot(html: bootHtml) { [weak self] bootResult in
            if let error = bootResult {
                self?.messenger?.setMessageHandlerOnChannel(self?.runtimeTxChannel(runtimeId) ?? "", binaryMessageHandler: nil)
                self?.runtimes.removeValue(forKey: runtimeId)?.close()
                result(["runtimeId": nil, "error": error] as [String: Any?])
            } else {
                NSLog("[RpcDartWasm] Runtime %@ booted on WKWebView", runtimeId)
                result(["runtimeId": runtimeId, "error": nil] as [String: Any?])
            }
        }
    }

    // MARK: - Close / byte channels

    private func closeRuntime(runtimeId: String) {
        messenger?.setMessageHandlerOnChannel(runtimeTxChannel(runtimeId), binaryMessageHandler: nil)
        runtimes.removeValue(forKey: runtimeId)?.close()
    }

    private func registerByteChannel(runtimeId: String) {
        messenger?.setMessageHandlerOnChannel(runtimeTxChannel(runtimeId), binaryMessageHandler: { [weak self] message, reply in
            guard let self = self, let bytes = message else {
                reply(nil)
                return
            }
            self.receiveRuntimeBytes(runtimeId: runtimeId, bytes: bytes)
            reply(nil)
        })
    }

    private func receiveRuntimeBytes(runtimeId: String, bytes: Data) {
        guard let runtime = runtimes[runtimeId] else { return }
        runtime.enqueueForJS(bytes)
    }

    private func stripModuleSyntax(_ code: String) -> String {
        return code
            .replacingOccurrences(of: "export async function ", with: "async function ")
            .replacingOccurrences(of: "export const ", with: "const ")
            .replacingOccurrences(of: "export function ", with: "function ")
            .replacingOccurrences(of: "export class ", with: "class ")
    }

    private func runtimeTxChannel(_ runtimeId: String) -> String {
        return "rpc_dart_wasm/\(runtimeId)/outgoing"
    }

    private func runtimeRxChannel(_ runtimeId: String) -> String {
        return "rpc_dart_wasm/\(runtimeId)/incoming"
    }
}

// MARK: - SchemeHandler (zero-copy byte transport + WASM serving)

private class SchemeHandler: NSObject, WKURLSchemeHandler {
    private var wasmBytes: Data?
    private var pendingRecvTask: WKURLSchemeTask?
    private var recvQueue: [Data] = []
    private var stopped = false
    var onSend: ((Data) -> Void)?

    init(wasmBytes: Data) {
        self.wasmBytes = wasmBytes
        super.init()
    }

    func enqueue(_ data: Data) {
        if let task = pendingRecvTask {
            pendingRecvTask = nil
            respond(task: task, data: data)
        } else {
            recvQueue.append(data)
        }
    }

    func stop() {
        stopped = true
        wasmBytes = nil
        if let task = pendingRecvTask {
            pendingRecvTask = nil
            let response = HTTPURLResponse(
                url: task.request.url!,
                statusCode: 499,
                httpVersion: nil,
                headerFields: nil
            )!
            task.didReceive(response)
            task.didFinish()
        }
        recvQueue.removeAll()
    }

    func webView(_ webView: WKWebView, start urlSchemeTask: WKURLSchemeTask) {
        guard !stopped else {
            urlSchemeTask.didFailWithError(NSError(domain: "rpc-wasm", code: -1))
            return
        }

        let path = urlSchemeTask.request.url?.path ?? ""

        if path == "/module.wasm" || path == "module.wasm" {
            guard let wasm = wasmBytes else {
                urlSchemeTask.didFailWithError(NSError(domain: "rpc-wasm", code: -2))
                return
            }
            let response = HTTPURLResponse(
                url: urlSchemeTask.request.url!,
                statusCode: 200,
                httpVersion: nil,
                headerFields: [
                    "Content-Type": "application/wasm",
                    "Content-Length": "\(wasm.count)"
                ]
            )!
            urlSchemeTask.didReceive(response)
            urlSchemeTask.didReceive(wasm)
            urlSchemeTask.didFinish()
            self.wasmBytes = nil
            return
        }

        if path == "/send" || path == "send" {
            let bodyData = readBody(from: urlSchemeTask.request)
            if !bodyData.isEmpty {
                onSend?(bodyData)
            }
            let response = HTTPURLResponse(
                url: urlSchemeTask.request.url!,
                statusCode: 204,
                httpVersion: nil,
                headerFields: nil
            )!
            urlSchemeTask.didReceive(response)
            urlSchemeTask.didFinish()
            return
        }

        if path == "/recv" || path == "recv" {
            if !recvQueue.isEmpty {
                let data = recvQueue.removeFirst()
                respond(task: urlSchemeTask, data: data)
            } else {
                if let prev = pendingRecvTask {
                    prev.didFailWithError(NSError(domain: "rpc-wasm", code: -5,
                        userInfo: [NSLocalizedDescriptionKey: "superseded by new recv"]))
                }
                pendingRecvTask = urlSchemeTask
            }
            return
        }

        urlSchemeTask.didFailWithError(NSError(domain: "rpc-wasm", code: -3))
    }

    func webView(_ webView: WKWebView, stop urlSchemeTask: WKURLSchemeTask) {
        if pendingRecvTask === urlSchemeTask {
            pendingRecvTask = nil
        }
    }

    private func readBody(from request: URLRequest) -> Data {
        if let body = request.httpBody, !body.isEmpty {
            return body
        }
        guard let stream = request.httpBodyStream else { return Data() }
        stream.open()
        defer { stream.close() }
        var result = Data()
        let bufferSize = 16384
        var buffer = [UInt8](repeating: 0, count: bufferSize)
        while stream.hasBytesAvailable {
            let read = stream.read(&buffer, maxLength: bufferSize)
            if read > 0 {
                result.append(buffer, count: read)
            } else {
                break
            }
        }
        return result
    }

    private func respond(task: WKURLSchemeTask, data: Data) {
        let response = HTTPURLResponse(
            url: task.request.url!,
            statusCode: 200,
            httpVersion: nil,
            headerFields: [
                "Content-Type": "application/octet-stream",
                "Content-Length": "\(data.count)"
            ]
        )!
        task.didReceive(response)
        task.didReceive(data)
        task.didFinish()
    }
}

// MARK: - WasmRuntime (WKWebView wrapper)

private class WasmRuntime: NSObject, WKScriptMessageHandler, WKNavigationDelegate {
    let runtimeId: String
    let webView: WKWebView
    private let messenger: FlutterBinaryMessenger
    private let rxChannel: String
    private let schemeHandler: SchemeHandler
    private var bootCompletion: ((String?) -> Void)?
    private var bootTimeoutTimer: Timer?

    private static let bootTimeoutSeconds: TimeInterval = 30

    init(runtimeId: String, messenger: FlutterBinaryMessenger, rxChannel: String, wasmBytes: Data) {
        self.runtimeId = runtimeId
        self.messenger = messenger
        self.rxChannel = rxChannel
        self.schemeHandler = SchemeHandler(wasmBytes: wasmBytes)

        let config = WKWebViewConfiguration()
        config.setURLSchemeHandler(schemeHandler, forURLScheme: "rpc-wasm")
        self.webView = WKWebView(frame: .zero, configuration: config)
        super.init()

        schemeHandler.onSend = { [weak self] data in
            guard let self = self else { return }
            self.messenger.send(onChannel: self.rxChannel, message: data)
        }

        webView.navigationDelegate = self
        webView.configuration.userContentController.add(self, name: "rpcBoot")
        webView.configuration.userContentController.add(self, name: "rpcConsole")
    }

    func boot(html: String, completion: @escaping (String?) -> Void) {
        bootCompletion = completion
        webView.loadHTMLString(html, baseURL: nil)

        bootTimeoutTimer = Timer.scheduledTimer(withTimeInterval: Self.bootTimeoutSeconds, repeats: false) { [weak self] _ in
            guard let self = self, let cb = self.bootCompletion else { return }
            self.bootCompletion = nil
            self.bootTimeoutTimer = nil
            cb("boot_timeout: no response within \(Int(Self.bootTimeoutSeconds))s")
        }
    }

    func enqueueForJS(_ data: Data) {
        schemeHandler.enqueue(data)
    }

    func close() {
        bootTimeoutTimer?.invalidate()
        bootTimeoutTimer = nil
        schemeHandler.stop()
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "rpcBoot")
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "rpcConsole")
        webView.stopLoading()
        webView.loadHTMLString("", baseURL: nil)
    }

    private func finishBoot(_ error: String?) {
        bootTimeoutTimer?.invalidate()
        bootTimeoutTimer = nil
        let cb = bootCompletion
        bootCompletion = nil
        cb?(error)
    }

    // MARK: - WKNavigationDelegate

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        finishBoot("navigation_failed: \(error.localizedDescription)")
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        finishBoot("navigation_failed: \(error.localizedDescription)")
    }

    func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
        finishBoot("content_process_terminated")
    }

    // MARK: - WKScriptMessageHandler

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == "rpcBoot" {
            let body = message.body as? String ?? ""
            if body == "ok" {
                finishBoot(nil)
            } else {
                let error = body.hasPrefix("error:") ? String(body.dropFirst(6)) : body
                finishBoot(error)
            }
            return
        }

        if message.name == "rpcConsole" {
            if let dict = message.body as? [String: Any],
               let level = dict["level"] as? String,
               let msg = dict["message"] as? String {
                let prefix: String
                switch level {
                case "warn": prefix = "W"
                case "error": prefix = "E"
                case "debug": prefix = "D"
                default: prefix = "I"
                }
                let log = "\(prefix):\(msg)"
                let consoleChannel = "rpc_dart_wasm/\(runtimeId)/console"
                if let data = log.data(using: .utf8) {
                    messenger.send(onChannel: consoleChannel, message: data)
                }
            }
            return
        }

    }
}

File: packages/transport/rpc_dart_wasm/android/src/main/kotlin/com/nogipx/rpc_dart_wasm/RpcDartWasmPlugin.kt
package com.nogipx.rpc_dart_wasm

import android.content.Context
import androidx.javascriptengine.JavaScriptIsolate
import androidx.javascriptengine.JavaScriptSandbox
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.guava.await
import kotlinx.coroutines.launch
import java.nio.ByteBuffer
import java.util.UUID

class RpcDartWasmPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var messenger: BinaryMessenger
    private lateinit var context: Context
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private val runtimes = mutableMapOf<String, JavaScriptIsolate>()
    private var sandbox: JavaScriptSandbox? = null
    private var sandboxFuture: kotlinx.coroutines.Deferred<JavaScriptSandbox>? = null
    private var messageCounter = 0

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "rpc_dart_wasm")
        channel.setMethodCallHandler(this)
        messenger = binding.binaryMessenger
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.launch {
            runtimes.keys.forEach { runtimeId ->
                messenger.setMessageHandler(runtimeTxChannel(runtimeId), null)
            }
            runtimes.values.forEach { it.close() }
            runtimes.clear()
            sandbox?.close()
            sandbox = null
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        scope.launch {
            try {
                when (call.method) {
                    "checkSupport" -> result.success(checkSupport())
                    "loadRuntime" -> result.success(loadRuntime(
                        call.argument("wasm"),
                        call.argument("mjs"),
                        call.argument<String>("jsBootPrefix") ?: "",
                    ))
                    "closeRuntime" -> {
                        closeRuntime(call.argument<String>("runtimeId")!!)
                        result.success(null)
                    }
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                android.util.Log.e("RpcDartWasm", "Error in ${call.method}", e)
                result.error("RPC_WASM_ERROR", e.message, e.stackTraceToString())
            }
        }
    }

    private suspend fun ensureSandbox(): JavaScriptSandbox {
        if (sandbox != null) return sandbox!!
        if (sandboxFuture == null) {
            sandboxFuture = scope.async {
                JavaScriptSandbox.createConnectedInstanceAsync(context).await()
            }
        }
        sandbox = sandboxFuture!!.await()
        return sandbox!!
    }

    private suspend fun checkSupport(): Map<String, Any> {
        val results = mutableMapOf<String, Any>()
        val supported = JavaScriptSandbox.isSupported()
        results["jsEngineAvailable"] = supported
        if (!supported) return results

        val sb = ensureSandbox()
        results["wasmCompilationSupported"] =
            sb.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_WASM_COMPILATION)
        results["namedDataSupported"] =
            sb.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_PROVIDE_CONSUME_ARRAY_BUFFER)

        val isolate = sb.createIsolate()
        try {
            val hasWasm = isolate.evaluateJavaScriptAsync(
                "(typeof WebAssembly !== 'undefined').toString()"
            ).await()
            results["hasWebAssembly"] = hasWasm == "true"
            if (hasWasm == "true") {
                val gcCheck = isolate.evaluateJavaScriptAsync("""
                    (function() {
                      try {
                        var bytes = new Uint8Array([
                          0x00, 0x61, 0x73, 0x6D, 0x01, 0x00, 0x00, 0x00,
                          0x01, 0x05, 0x01, 0x5F, 0x01, 0x7F, 0x01
                        ]);
                        return WebAssembly.validate(bytes) ? 'GC_SUPPORTED' : 'GC_NOT_VALIDATED';
                      } catch(e) { return 'GC_ERROR: ' + e; }
                    })()
                """.trimIndent()).await()
                results["wasmGC"] = gcCheck
            }
        } finally {
            isolate.close()
        }
        return results
    }

    private suspend fun loadRuntime(
        wasmBytes: ByteArray?,
        mjsCode: String?,
        jsBootPrefix: String,
    ): Map<String, Any?> {
        if (wasmBytes == null || mjsCode == null) {
            return mapOf("error" to "Missing wasm or mjs data")
        }

        val sb = ensureSandbox()
        if (!sb.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_WASM_COMPILATION)) {
            return mapOf("runtimeId" to null, "error" to "JS_FEATURE_WASM_COMPILATION is not supported")
        }
        if (!sb.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_PROVIDE_CONSUME_ARRAY_BUFFER)) {
            return mapOf("error" to "Named data API not supported")
        }

        val runtimeId = UUID.randomUUID().toString()
        val isolate = sb.createIsolate()
        runtimes[runtimeId] = isolate
        registerByteChannel(runtimeId)
        isolate.provideNamedData("rpc_wasm_module", wasmBytes)

        val plainJs = stripModuleSyntax(mjsCode)
        val bootScript = """
            var globalThis = this;
            var _rpcWasmOutbox = [];
            var _microtaskQueue = [];
            var _rpcWasmBootError = null;
            var _rpcWasmBootTrace = null;
            var _rpcWasmBootPhase = 'init';
            var _timerId = 0;
            var _timers = {};
            function queueMicrotask(fn) { _microtaskQueue.push(fn); }
            function _flushMicrotasks() {
              while (_microtaskQueue.length > 0) {
                var fn = _microtaskQueue.shift();
                try { fn(); } catch(e) { console.error(e); }
              }
            }
            var _rpcConsoleLog = [];
            var console = {
              log: function() { _rpcConsoleLog.push('I:' + Array.prototype.join.call(arguments, ' ')); },
              warn: function() { _rpcConsoleLog.push('W:' + Array.prototype.join.call(arguments, ' ')); },
              error: function() { _rpcConsoleLog.push('E:' + Array.prototype.join.call(arguments, ' ')); },
              info: function() { _rpcConsoleLog.push('I:' + Array.prototype.join.call(arguments, ' ')); },
              debug: function() { _rpcConsoleLog.push('D:' + Array.prototype.join.call(arguments, ' ')); }
            };
            function _rpcDrainConsole() {
              if (_rpcConsoleLog.length === 0) return '';
              var out = _rpcConsoleLog.join('\n');
              _rpcConsoleLog = [];
              return out;
            }
            function setTimeout(fn, ms) {
              var id = ++_timerId;
              _timers[id] = { fn: fn, interval: false, ms: ms || 0, next: Date.now() + (ms || 0) };
              return id;
            }
            function setInterval(fn, ms) {
              var id = ++_timerId;
              _timers[id] = { fn: fn, interval: true, ms: ms || 16, next: Date.now() + (ms || 16) };
              return id;
            }
            function clearInterval(id) { delete _timers[id]; }
            function clearTimeout(id) { delete _timers[id]; }
            function _tickTimers() {
              var now = Date.now();
              var ids = Object.keys(_timers);
              var active = 0;
              for (var i = 0; i < ids.length; i++) {
                var id = ids[i];
                var t = _timers[id];
                if (!t) continue;
                if (now >= t.next) {
                  try { t.fn(); } catch(e) { console.error(e); }
                  if (t.interval) {
                    t.next = now + t.ms;
                    active++;
                  } else {
                    delete _timers[id];
                  }
                } else {
                  active++;
                }
              }
              _flushMicrotasks();
              return '' + active;
            }
            var _b64c = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
            function _bytesToBase64(bytes) {
              var r = '', i = 0, len = bytes.length;
              while (i < len) {
                var a = bytes[i++], b = i < len ? bytes[i++] : -1, c = i < len ? bytes[i++] : -1;
                r += _b64c[a >> 2];
                if (b < 0) { r += _b64c[(a & 3) << 4] + '=='; }
                else if (c < 0) { r += _b64c[((a & 3) << 4) | (b >> 4)] + _b64c[(b & 15) << 2] + '='; }
                else { r += _b64c[((a & 3) << 4) | (b >> 4)] + _b64c[((b & 15) << 2) | (c >> 6)] + _b64c[c & 63]; }
              }
              return r;
            }
            function _rpcWasmSendBytes(bytes) {
              _rpcWasmOutbox.push(_bytesToBase64(bytes));
            }
            function _rpcWasmDrainOutbox() {
              if (_rpcWasmOutbox.length === 0) return '';
              var out = _rpcWasmOutbox.join('\n');
              _rpcWasmOutbox = [];
              return out;
            }
            function _rpcWasmReceiveBytes(bytes) {
              if (typeof rpcWasmReceiveBytes === 'function') {
                rpcWasmReceiveBytes(bytes);
              } else if (typeof globalThis.rpcWasmReceiveBytes === 'function') {
                globalThis.rpcWasmReceiveBytes(bytes);
              }
              _flushMicrotasks();
            }
            $jsBootPrefix
            $plainJs
            (async function() {
              try {
                if (typeof WebAssembly === 'undefined') {
                  throw new Error('WebAssembly is not available in this JS runtime');
                }
                _rpcWasmBootPhase = 'consumeNamedData';
                var wasmBuf = await android.consumeNamedDataAsArrayBuffer('rpc_wasm_module');
                var wasmBytes = new Uint8Array(wasmBuf);
                _rpcWasmBootPhase = 'compile';
                var compiled = await compile(wasmBytes);
                _rpcWasmBootPhase = 'instantiate';
                var app = await compiled.instantiate({});
                _rpcWasmBootPhase = 'invokeMain';
                app.invokeMain();
                _rpcWasmBootPhase = 'flushMicrotasks';
                _flushMicrotasks();
                _rpcWasmBootPhase = 'done';
                return 'ok';
              } catch (e) {
                _rpcWasmBootError = '' + e;
                _rpcWasmBootTrace = e && e.stack ? '' + e.stack : null;
                throw e;
              }
            })()
        """.trimIndent()

        return try {
            val evalResult = isolate.evaluateJavaScriptAsync(bootScript).await()
            drainAndPush(runtimeId)
            android.util.Log.d("RpcDartWasm", "Runtime $runtimeId booted: ${evalResult.take(100)}")
            mapOf("runtimeId" to runtimeId, "error" to null)
        } catch (e: Exception) {
            val jsError = runCatching {
                isolate.evaluateJavaScriptAsync(
                    "JSON.stringify({ error: _rpcWasmBootError, trace: _rpcWasmBootTrace, phase: _rpcWasmBootPhase })"
                ).await()
            }.getOrNull()
            android.util.Log.e(
                "RpcDartWasm",
                "Runtime boot failed: $runtimeId jsError=$jsError",
                e,
            )
            mapOf(
                "runtimeId" to runtimeId,
                "error" to (jsError ?: e.message ?: e.toString()),
            )
        }
    }

    private suspend fun forwardBytesToRuntime(runtimeId: String, bytes: ByteArray) {
        val isolate = runtimes[runtimeId] ?: return
        android.util.Log.d(
            "RpcDartWasm",
            "Forwarding ${bytes.size} bytes to runtime $runtimeId",
        )
        val name = "rpc_msg_${messageCounter++}"
        isolate.provideNamedData(name, bytes)
        isolate.evaluateJavaScriptAsync("""
            (async function() {
              var buf = await android.consumeNamedDataAsArrayBuffer('$name');
              _rpcWasmReceiveBytes(new Uint8Array(buf));
              return 'ok';
            })()
        """.trimIndent()).await()
        drainAndPush(runtimeId)
    }

    private suspend fun drainAndPush(runtimeId: String) {
        val isolate = runtimes[runtimeId] ?: return

        // Drain console logs and forward to Flutter.
        val consoleLogs = isolate.evaluateJavaScriptAsync("_rpcDrainConsole()").await()
        if (consoleLogs.isNotEmpty()) {
            sendConsoleLog(runtimeId, consoleLogs)
        }

        val raw = isolate.evaluateJavaScriptAsync("_rpcWasmDrainOutbox()").await()
        if (raw.isEmpty()) return

        for (b64 in raw.split('\n')) {
            if (b64.isEmpty()) continue
            val bytes = android.util.Base64.decode(b64, android.util.Base64.DEFAULT)
            sendRuntimeBytes(runtimeId, bytes)
        }
    }

    private fun sendConsoleLog(runtimeId: String, log: String) {
        val channel = "rpc_dart_wasm/$runtimeId/console"
        val bytes = log.toByteArray(Charsets.UTF_8)
        val buffer = java.nio.ByteBuffer.allocateDirect(bytes.size)
        buffer.put(bytes)
        messenger.send(channel, buffer)
    }

    private fun closeRuntime(runtimeId: String) {
        messenger.setMessageHandler(runtimeTxChannel(runtimeId), null)
        try {
            runtimes.remove(runtimeId)?.close()
        } catch (e: Exception) {
            android.util.Log.w("RpcDartWasm", "Isolate close for $runtimeId: ${e.message}")
        }
    }

    private fun registerByteChannel(runtimeId: String) {
        messenger.setMessageHandler(runtimeTxChannel(runtimeId)) { message, reply ->
            val bytes = message?.let { buffer ->
                val copy = buffer.asReadOnlyBuffer()
                copy.rewind()
                val data = ByteArray(copy.remaining())
                copy.get(data)
                data
            }

            scope.launch {
                try {
                    if (bytes != null) {
                        forwardBytesToRuntime(runtimeId, bytes)
                    }
                } finally {
                    reply.reply(null)
                }
            }
        }
    }

    private fun sendRuntimeBytes(runtimeId: String, bytes: ByteArray) {
        android.util.Log.d(
            "RpcDartWasm",
            "Sending ${bytes.size} bytes back to Flutter from runtime $runtimeId",
        )
        // Flutter JNI uses buffer.position() as data length, NOT as read offset.
        // Do NOT call flip() or rewind() — leave position at end after put().
        val buffer = ByteBuffer.allocateDirect(bytes.size)
        buffer.put(bytes)
        messenger.send(runtimeRxChannel(runtimeId), buffer)
    }

    private fun runtimeTxChannel(runtimeId: String) = "rpc_dart_wasm/$runtimeId/outgoing"

    private fun runtimeRxChannel(runtimeId: String) = "rpc_dart_wasm/$runtimeId/incoming"

    private fun stripModuleSyntax(code: String): String = code
        .replace("export async function ", "async function ")
        .replace("export const ", "const ")
        .replace("export function ", "function ")
        .replace("export class ", "class ")
}
